### DISCUSSION TITLE
TDI's Indicator Buffer Sequence

### DISCUSSION THREAD

**Discussion by mavidelisi** [EA31337/Strategy-TDI#4] [General]
```
  SetIndexBuffer(0, gdaVolaTop);		//TDI_UP_LINE
  SetIndexBuffer(1, gdaVolaBtm);		//TDI_DN_LINE
  SetIndexBuffer(2, gdaMktBase);		//TDI_MID_LINE
  SetIndexBuffer(3, gdaTradeSig);		//TDI_TRADE_SIGNAL_LINE
  SetIndexBuffer(4, gdaRSISig); 		//TDI_RSI_LINE
  SetIndexBuffer(5, gdaRSI);      		//RSI MA Price Line
  IndicatorDigits(1);
  IndicatorShortName(INDICATOR_NAME);
  
  
  // Indicator line identifiers used in the indicator.
enum ENUM_TDI_MODE {
  TDI_RSI_LINE = 0,       // RSI line
  TDI_UP_LINE,            // VB High
  TDI_MID_LINE,           // Base line
  TDI_DN_LINE,            // VB Low
  TDI_RSI_MA_LINE,        // RSI MA Price Line
  TDI_TRADE_SIGNAL_LINE,  // Trade Signal Line
  FINAL_TDI_MODE_ENTRY,
};
```

Above. 1. paragraph buffer taken from the inside of the indicator, 2. the paragraph is Strategy-TDI/Indi_TDI.mqh the  values that are a reference to the buffer in the indicator. But their ranks are different.

Isn't the sequence number important here ? So according to the buffer values in the indicator, shouldn't it be in the order as follows, or am I wrong? Can you clarify a little?

```
enum ENUM_TDI_MODE {
  TDI_UP_LINE=0,          // VB High
  TDI_DN_LINE,            // VB Low
  TDI_MID_LINE,           // Base line
  TDI_TRADE_SIGNAL_LINE,  // Trade Signal Line
  TDI_RSI_LINE = 0,       // RSI line
  TDI_RSI_MA_LINE,        // RSI MA Price Line
  FINAL_TDI_MODE_ENTRY,
};
```

**Comment by kenorb** [comment]
Yes, the sequence is important.

Currently `Indi_TDI.mqh` is implementing the original version of TDI indicator ([`TDI.mq4`](https://github.com/EA31337/EA31337-indicators-other/blob/76f02848b7a5a180a11645fb346044cfcc0a2614/Oscillators/Multi/TDI.mq4)), not the [`TDI-RT-Clone.mq4`](https://github.com/EA31337/EA31337-indicators-other/blob/e0ba0f914fc59538830929b0d6ae7bf5fd398829/Oscillators/Multi/TDI-RT-Clone.mq4). These, it seems have different buffer sequence. If you'd like to use cloned version, you'll have to change the enum items accordingly.

Let's break it down the buffers from [`TDI.mq4`](https://github.com/EA31337/EA31337-indicators-other/blob/76f02848b7a5a180a11645fb346044cfcc0a2614/Oscillators/Multi/TDI.mq4):

```
   SetIndexBuffer(0,RSIBuf);
   SetIndexBuffer(1,UpZone);
   SetIndexBuffer(2,MdZone);
   SetIndexBuffer(3,DnZone);
   SetIndexBuffer(4,MaBuf);
   SetIndexBuffer(5,MbBuf);
...
   SetIndexLabel(0,NULL);
   SetIndexLabel(1,"VB High");
   SetIndexLabel(2,"Market Base Line");
   SetIndexLabel(3,"VB Low");
   SetIndexLabel(4,"RSI Price Line");
   SetIndexLabel(5,"Trade Signal Line");
```

And with these colors:

```
#property indicator_color1 Black
#property indicator_color2 0XFFFFFFFF  // MediumBlue
#property indicator_color3 0xFFFFFFFF  // Yellow
#property indicator_color4 0XFFFFFFFF  // MediumBlue
#property indicator_color5 Green
#property indicator_color6 Red
```

and description:

```
Green line = RSI Price line
Red line = Trade Signal line
Blue lines = Volatility Band
Yellow line = Market Base Line
```

To summarize the buffers based on the above code:

0. RSIBuf, NULL, Black
1. UpZone, VB High, MediumBlue, Blue lines = Volatility Band
2. MdZone, Market Base Line, Yellow, Yellow line = Market Base Line
3. DnZone, VB Low, MediumBlue, Blue lines = Volatility Band
4. MaBuf, RSI Price Line, Green, Green line = RSI Price line
5. MbBuf, Trade Signal Line, Red, Red line = Trade Signal line

So the enum defined in `Indi_TDI.mqh` is the correct one:

```
// Indicator line identifiers used in the indicator.
enum ENUM_TDI_MODE {
  TDI_RSI_LINE = 0,       // RSI line
  TDI_UP_LINE,            // VB High
  TDI_MID_LINE,           // Base line
  TDI_DN_LINE,            // VB Low
  TDI_RSI_MA_LINE,        // RSI MA Price Line
  TDI_TRADE_SIGNAL_LINE,  // Trade Signal Line
  FINAL_TDI_MODE_ENTRY,
};
```

For the cloned version of TDI indicator, you'll have to do adjustments.

**Reply by mavidelisi** [reply]
Ok. My mistake. Thanks.

**Reply by kenorb** [reply]
To handle `TDI-RT-Clone` indicator, I've added another separate class for it (`Indi_TDI_RT`) with the following enum items:

```
enum ENUM_TDI_RT_MODE {
  TDI_RT_VOLA_TOP = 0,       // Volatility Top
  TDI_RT_VOLA_BTM,           // Volatility Bottom
  TDI_RT_MARKET_BASE,        // Market Base
  TDI_RT_TRADE_SIGNAL_LINE,  // Trade Signal Line
  TDI_RT_RSI_SIGNAL_LINE,    // RSI Signal Line
  TDI_RT_RSI_LINE,           // RSI line
  FINAL_TDI_RT_MODE_ENTRY,
};
```

The TDI strategy can use both indicators at the same time by allowing users to select which type of indicator they want to run (so the results can be compared against). I'll add such option later on.

Commit: 01d788431ffecbfb48d482c888b9ca6cd0782be7.
